### PR TITLE
feat!: initial implementation of parser

### PR DESCRIPTION
**Pull request by bcoe**
This is the start of a parser, based on a formal grammar for conventional commits (also in this repository).

After discussion with @byCedric, I'm loosely basing the parse tree on [uninst](https://github.com/syntax-tree/unist#data). I would love help with this effort.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+/*
+ * <ZWNBSP>       ::= "U+FEFF"
+ * <TAB>          ::= "U+0009"
+ * <VT>           ::= "U+000B"
+ * <FF>           ::= "U+000C"
+ * <SP>           ::= "U+0020"
+ * <NBSP>         ::= "U+00A0"
+*/
+function isWhitespace (token) {
+  return token === '\ufeff' || token === '\u0009' || token === '\u000b' || token === '\u000c' || token === '\u0020' || token === '\u00a0'
```
Comment (byCedric) [inline index.js:55]: We could make these tokens a bit more readable if we use something like `codes.tab`, [like micromark](https://github.com/micromark/micromark/blob/main/lib/character/codes.mjs)

File: test.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+const assert = require('assert')
+const { describe, it } = require('mocha')
+const parser = require('./')
+
+describe('parser', () => {
+  describe('<type> ":" <text>', () => {
+    it('parses summary with no scope', () => {
+      const message = parser('fix: a really weird bug')
+      assert.strictEqual(message.type, 'message')
```
Comment (byCedric) [inline test.js:9]: We could simplify this if we can match on either a deep object or json string, in combination with [`unist-builder`](https://www.npmjs.com/package/unist-builder).

File: test.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      const summary = message.children[0]
+      assert.strictEqual(summary.type, 'summary')
+      const type = summary.children[0]
+      assert.strictEqual(type.type, 'type')
+      assert.strictEqual(type.value, 'fix')
```
Comment (bcoe) [inline test.js:14]: should we consider using snapshot testing, or is it better to be explicit about the nodes?

**Review by byCedric — COMMENTED** [review]
Looks awesome! I'll test run it this weekend too. One thing that could help during development/debugging is adding `debug` messages. With that, the code path should be visible. (e.g. [this](https://github.com/micromark/micromark/blob/ac44b027357e36694efd2c59babba1b89515e73c/lib/util/create-tokenizer.mjs#L109))

File: test.js (hunk 3 above)
Comment (byCedric) [reply test.js:14]: Those snapshots won't change that often I think. Either way looks fine to me.

File: README.md [hunk 4]
```diff
@@ -26,20 +39,20 @@ Reference implementation of Conventional Commits specification.
...
-<value>        ::= <summary-text> 1*<continuation>
-                |  <summary-text>
-<continuation> ::= <newline> <whitespace> <summary-text>
+<value>        ::= <text> 1*<continuation>
+                |  <text>
+<continuation> ::= <newline> <whitespace> <text>
 
 <body>         ::= 1*<body-text>
-<body-text>    ::= <newline>? <summary-text>
+<body-text>    ::= <newline>? <text>
```
Comment (bcoe) [inline README.md:57]: This isn't quite right, as we'd like to allow `\n\n` in message bodies, I  think perhaps we should allow `\n\n` unless

The next token matches a `<footer>`.

File: test.js (hunk 2 above)
Comment (bcoe) [reply test.js:9]: I think at that point using snapshot testing [like in c8](https://github.com/bcoe/c8/blob/master/test/integration.js#L31) would perhaps be the most elegant, I expect the trees will be fairly readable.

**Pull request merged**